[iOS] Rewrite the guide to integrate with Brownfield apps (#4239)
facebook/react-native-website: docs/_integration-with-existing-apps-ios.md

import Tabs from '@theme/Tabs'; import TabItem from '@theme/TabItem'; import constants from '@site/core/TabsConstants';

## Key Concepts

The keys to integrating React Native components into your iOS application are to:

1. Set up the correct directory structure.
2. Install the necessary NPM dependencies.
3. Adding React Native to your Podfile configuration.
4. Writing the TypeScript code for your first React Native screen.
5. Integrate React Native with your iOS code using a `RCTRootView`.
6. Testing your integration by running the bundler and seeing your app in action.

## Using the Community Template

> While you follow this guide, we suggest you to use the [React Native Community Template](https://github.com/react-native-community/template/) as reference. The template contains a **minimal Android app** and will help you understanding how to integrate React Native into an existing Android app.

Reviewer: ```suggestion
While you follow this guide, we suggest you to use the [React Native Community Template](https://github.com/react-native-community/template/) as reference. The template contains a **minimal iOS app** and will help you understanding how to integrate React Native into an existing iOS app.
```

## Prerequisites

Follow the guide on [setting up your development environment](set-up-your-environment) and using [React Native without a framework](getting-started-without-a-framework) to configure your development environment for building React Native apps for iOS.
This guide also assumes you're familiar with the basics of iOS development such as creating a `UIViewController` and editing the `Podfile` file.

### 1. Set up directory structure

To ensure a smooth experience, create a new folder for your integrated React Native project, then **move your existing iOS project** to the `/ios` subfolder.

## 2. Install NPM dependencies

Go to the root directory and run the following command:

```

> wget https://raw.githubusercontent.com/react-native-community/template/refs/heads/0.76-stable/template/package.json

Reviewer: Is it possible to move this `0.76-stable` into a constant? This could make future maintenance easier

Author: no idea, honestly.... We should look into it.

```

This will copy the `package.json` [file from the Community template](https://github.com/react-native-community/template/blob/0.76-stable/template/package.json) to your project.

Next, install the NPM packages by running:

<Tabs groupId="package-manager" queryString defaultValue={constants.defaultPackageManager} values={constants.packageManagers}>
<TabItem value="npm">

```shell
npm install
```

</TabItem>
<TabItem value="yarn">

```shell
yarn install
```

</TabItem>
</Tabs>

Installation process has created a new `node_modules` folder. This folder stores all the JavaScript dependencies required to build your project.

> Add `node_modules/` to your `.gitignore` file (here the [Community default one](https://github.com/react-native-community/template/blob/0.76-stable/template/_gitignore)).

Reviewer: nit: We can also suggest to use wget here

Author: I thought about it, but they might not be using git as source control tool.

### 3. Install Development tools

### Command Line Tools for Xcode

Install the Command Line Tools. Choose **Settings... (or Preferences...)** in the Xcode menu. Go to the Locations panel and install the tools by selecting the most recent version in the Command Line Tools dropdown.

![Xcode Command Line Tools](/docs/assets/GettingStartedXcodeCommandLineTools.png)

### CocoaPods

> [CocoaPods](https://cocoapods.org) is a package management tool for iOS and macOS development. We use it to add the actual React Native framework code locally into your current project.

Reviewer: I'm not sure about this sentence:

> We use it to add the actual React Native framework code locally into your current project.

Maybe something like: We use it to integrate React Native's code and dependencies into your current project.

We recommend installing CocoaPods using [Homebrew](https://brew.sh/):

```shell
brew install cocoapods
```

## 4. Adding React Native to your app

### Configuring CocoaPods

To configure CocoaPods, we need two files:

- A **Gemfile** that defines which Ruby dependencies we need.
- A **Podfile** that defines how to properly install our dependencies.

For the **Gemfile**, go to the root directory of your project and run this command

```sh
wget https://raw.githubusercontent.com/react-native-community/template/refs/heads/0.76-stable/template/Gemfile

> ```

Reviewer: This could be marked as optional

Author: Yes, but there are som many issues that can occur if you don't use the right gems that I'd rather keep it.

bundle exec pod install
```

The first command will install the ruby dependencies and the second command will actually integrate the React Native code in your application so that your iOS files can import the React Native headers.

## 5. Writing the TypeScript Code

Now we will actually modify the native Android application to integrate React Native.

The first bit of code we will write is the actual React Native code for the new screen that will be integrated into our application.

### Create a `index.js` file

First, create an empty `index.js` file in the root of your React Native project.

`index.js` is the starting point for React Native applications, and it is always required. It can be a small file that `import`s other file that are part of your React Native component or application, or it can contain all the code that is needed for it.

Our `index.js` should look as follows (here the [Community template file as reference](https://github.com/react-native-community/template/blob/0.76-stable/template/index.js)):

```js
import {AppRegistry} from 'react-native';
import App from './App';

AppRegistry.registerComponent('HelloWorld', () => App);
```

### Create a `App.tsx` file

> Then, let's create a `App.tsx` file. This is a TypeScript file that will contain the React Native component that we will integrate into our Android application.

Reviewer: ```suggestion
Then, let's create a `App.tsx` file. This is a TypeScript file that will contain the React Native component that we will integrate into our iOS application.
```

        contentInsetAdjustmentBehavior="automatic"
        style={backgroundStyle}>
        <Header />
        <View
          style={{
            backgroundColor: isDarkMode
              ? Colors.black
              : Colors.white,
            padding: 24,
          }}>
          <Text style={styles.title}>Step One</Text>
          <Text>
            Edit <Text style={styles.bold}>App.tsx</Text> to
            change this screen and see your edits.
          </Text>
          <Text style={styles.title}>See your changes</Text>
          <ReloadInstructions />
          <Text style={styles.title}>Debug</Text>
          <DebugInstructions />
        </View>
      </ScrollView>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  title: {
    fontSize: 24,
    fontWeight: '600',
  },
  bold: {
    fontWeight: '700',
  },
});

export default App;
```

Here the [Community template file as reference](https://github.com/react-native-community/template/blob/0.76-stable/template/App.tsx)

## 5. Integrating with your iOS code

We now need to add some native code in order to start the React Native runtime and tell it to render our React components.

### Requirements

> React Native is supposed to work with the `AppDelegate`. Sadly it does not supports the `SceneDelegate` pattern as of today.

Reviewer: This sets a negative tone, I would rephrase this to: 

React Native integrates into your project by subclassing `RCTAppDelegate`. Currently `SceneDelegate` is not supported.

or something like this, hopefully, we will remove this soon 🤞

Reviewer: Also if the user wants he can pass `automaticallyLoadReactNativeWindow = NO;` and then init his view in SceneDelegate by referencing RCTRootViewFactory from AppDelegate so its not like SceneDelegate won't work at all.

Author: Interesting... let's remove this for now. We can integrate the `SceneDelegate` in a next iteration

@end
```

2. Open the `AppDelegate.mm` file and modify it as it follows (See the official template's [AppDelegate.mm](https://github.com/react-native-community/template/blob/0.76-stable/template/ios/HelloWorld/AppDelegate.mm) as reference

```diff title="AppDelegate.mm"
#import "AppDelegate.h"
#import "ViewController.h"
+#import <React/RCTBundleURLProvider.h>

@interface AppDelegate ()

@end

@implementation AppDelegate {
  UIWindow *window;
}

 - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
+ self.automaticallyLoadReactNativeWindow = NO;
+ return [super application:application didFinishLaunchingWithOptions:launchOptions];
   window = [UIWindow new];
   window.rootViewController = [ViewController new];
   [window makeKeyAndVisible];
   return YES;

 }

+- (NSURL *)sourceURLForBridge:(RCTBridge *)bridge
+{
+  return [self bundleURL];
+}

+- (NSURL *)bundleURL
+{
+#if DEBUG
+  return [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index"];
+#else
+  return [[NSBundle mainBundle] URLForResource:@"main" withExtension:@"jsbundle"];
+#endif
+}
 @end
```

Let's have a look at the code above:

> 1. We are inherithing from the `RCTAppDelegate` and we are calling the `applicatinon:didFinishLaunchingWithOptions` of the `RCTAppDelegate`. This delegates all the React Native initialization processes to the base class.

Reviewer: ```suggestion
1. We are inheriting from the `RCTAppDelegate` and we are calling the `application:didFinishLaunchingWithOptions` of the `RCTAppDelegate`. This delegates all the React Native initialization processes to the base class.
```

2. We are customizing the RCTAppDelegate by setting the `automaticallyLoadReactNativeWindow` to `NO`. This step instruct React Native that the app is handling the `UIWindow` and React Native should not worry about that.

> 3. The methods `sourceURLForBridge:` and `bundleURL` are used by the App to tell to React Native wher it can find the JS bundle that needs to be rendered. The `sourceURLForBridge:` is from the Old Architecture and you can see that it is deferring the decision to the `bundleURL` method, required by the New Architecture.

Reviewer: ```suggestion
3. The methods `sourceURLForBridge:` and `bundleURL` are used by the App to tell React Native where it can find the JS bundle that needs to be rendered. The `sourceURLForBridge:` is from the Old Architecture and you can see that it is deferring the decision to the `bundleURL` method, required by the New Architecture.
```

```diff title="AppDelegate.swift"
import UIKit
+import React_RCTAppDelegate
+import React_RCTAppDelegate

@main
-class AppDelegate: UIResponder, UIApplicationDelegate {
+class AppDelegate: RCTAppDelegate {

-  var window: UIWindow?

-  func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
+  override func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
    // Override point for customization after application launch.
+    self.automaticallyLoadReactNativeWindow = false
+    super.application(application, didFinishLaunchingWithOptions: launchOptions)
    window = UIWindow()
-    window?.rootViewController = ViewController()
-    window?.makeKeyAndVisible()
+    window.rootViewController = ViewController()
+    window.makeKeyAndVisible()
    return true
  }

+  override func sourceURL(for bridge: RCTBridge) -> URL? {
+    self.bundleURL()
+  }

+  override func bundleURL() -> URL? {

> +#if DEBUG

Reviewer: It might be worth mentioning that users should check Active Compilation Conditions. In theory this should be set but if its not then this may lead to some issues 

![image](https://github.com/user-attachments/assets/ef0bc957-b014-4a26-a06e-3c43e24a0a70)

Author: Debug and Release flavors are created by default by Xcode. The only case when that flag is not there is for a project that started as Objective-C and got migrated to Swift - so Xcode didn't add that automation. 
But that's an edge case that people working on migrated projects should have already handled.
I'd rather keep the guide minimal, if possible.

+    RCTBundleURLProvider.sharedSettings().jsBundleURL(forBundleRoot: "index")
+#else
+    Bundle.main.url(forResource: "main", withExtension: "jsbundle")
+#endif
+  }
}
```

Let's have a look at the code above:

> 1. We are inherithing from the `RCTAppDelegate` and we are calling the `applicatinon(_:didFinishLaunchingWithOptions:)` of the `RCTAppDelegate`. This delegates all the React Native initialization processes to the base class.

Reviewer: ```suggestion
1. We are inheriting from the `RCTAppDelegate` and we are calling the `application (_:didFinishLaunchingWithOptions:)` of the `RCTAppDelegate`. This delegates all the React Native initialization processes to the base class.
```

</TabItem>
</Tabs>

#### Presenting a React Native view in a rootViewController

Finally, we can present our React Native view. To do so, we need a new View Controller that can host a view in which we can load the JS content.

1. From Xcode, let's create a new `UIViewController` (Let's call it `ReactViewController`).
2. Have the Initial `ViewController` present the `ReactViewController`. There are several ways to do so, depending on your app. For this example, we assume that you have a button that presents React Native Modally.

<Tabs groupId="ios-language" queryString defaultValue={constants.defaultAppleLanguage} values={constants.appleLanguages}>
<TabItem value="objc">

```diff title="ViewController.m"
#import "ViewController.h"
+#import "ReactViewController.h"

@interface ViewController ()

@end

@implementation ViewController

 - (void)viewDidLoad {
   [super viewDidLoad];
   // Do any additional setup after loading the view.
   self.view.backgroundColor = UIColor.systemBackgroundColor;

> +  UIButton * button = [UIButton new];

Reviewer: ```suggestion
+  UIButton *button = [UIButton new];
```

Author: we have a linter that fixes the code in the markdown in the website. :wink:

+  [button setTitle:@"Open React Native" forState:UIControlStateNormal];
+  [button setTitleColor:UIColor.systemBlueColor forState:UIControlStateNormal];
+  [button setTitleColor:UIColor.blueColor forState:UIControlStateHighlighted];
+  [button addTarget:self action:@selector(presentReactNative) forControlEvents:UIControlEventTouchUpInside];
+  [self.view addSubview:button];

+  button.translatesAutoresizingMaskIntoConstraints = NO;
+  [NSLayoutConstraint activateConstraints:@[
+    [button.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
+    [button.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor],
+    [button.centerYAnchor constraintEqualToAnchor:self.view.centerYAnchor],
+    [button.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
+  ]];
 }

+- (void)presentReactNative
+{

> +  [self presentViewController:[ReactViewController new] animated:YES completion:nil];

Reviewer: Shouldn't we retain this ViewController as a property on the class?

I am not sure but I think I've run into some memory management issues with deallocation when doing it this way

Author: It depends on what the VC does. In this simple example it should not be required, but let add it for safety.

+    button.addAction(UIAction { [weak self] _ in
+      self?.present(ReactViewController(), animated: true)
+    }, for: .touchUpInside)
+    self.view.addSubview(button)
+
+    button.translatesAutoresizingMaskIntoConstraints = false
+    NSLayoutConstraint.activate([
+      button.leadingAnchor.constraint(equalTo: self.view.leadingAnchor),
+      button.trailingAnchor.constraint(equalTo: self.view.trailingAnchor),
+      button.centerXAnchor.constraint(equalTo: self.view.centerXAnchor),
+      button.centerYAnchor.constraint(equalTo: self.view.centerYAnchor),
+    ])
  }
}
```

</TabItem>
</Tabs>

3. Update the `ReactViewController` code as it follows:

<Tabs groupId="ios-language" queryString defaultValue={constants.defaultAppleLanguage} values={constants.appleLanguages}>
<TabItem value="objc">

```diff title="ReactViewController.m"
#import "ReactViewController.h"
+#import <React-RCTAppDelegate/RCTRootViewFactory.h>
+#import <React-RCTAppDelegate/RCTAppDelegate.h>

@interface ReactViewController ()

@end

@implementation ReactViewController

 - (void)viewDidLoad {
   [super viewDidLoad];
   // Do any additional setup after loading the view.

> +   RCTRootViewFactory * factory = ((RCTAppDelegate *)RCTSharedApplication().delegate).rootViewFactory;

Reviewer: ```suggestion
+   RCTRootViewFactory *factory = ((RCTAppDelegate *)RCTSharedApplication().delegate).rootViewFactory;
```

+   self.view = [factory viewWithModuleName:@"HelloWorld"];
 }

@end
```

</TabItem>
<TabItem value="swift">

```diff title="ReactViewController.swift"
import UIKit
+import React_RCTAppDelegate

class ReactViewController: UIViewController {

  override func viewDidLoad() {
    super.viewDidLoad()

> +    let factory = (RCTSharedApplication()?.delegate as? RCTAppDelegate)?.rootViewFactory;
> +    self.view = factory?.view(withModuleName: "HelloWorld");

Reviewer: ```suggestion
+    let factory = (RCTSharedApplication()?.delegate as? RCTAppDelegate)?.rootViewFactory
+    self.view = factory?.view(withModuleName: "HelloWorld")
```

  }
}
```

</TabItem>
</Tabs>

> 4. Make sure to disable the Sandbox scripting. To achieve this, in Xcode, click on your app, then on build settings. Filter for script and set the `User Script Sandboxing` to `NO`.

Reviewer: Might be worth explaining why this is needed

![Disable Sandboxing](/docs/assets/disable-sandboxing.png);

## 6. Test your integration

> You have now done all the basic steps to integrate React Native with your current application. Now we will start the [Metro bundler](https://metrobundler.dev/) to build the TypeScript code of your application and have the server running on localhost to serve it.

Author: we should update this in Android too.

First, you need to create a `metro.config.js` file in the root of your project as follows:

```js
const {getDefaultConfig} = require('@react-native/metro-config');
module.exports = getDefaultConfig(__dirname);
```

You can checkout the [metro.config.js file](https://github.com/react-native-community/template/blob/0.76-stable/template/metro.config.js) from the Community template file as reference.

Once you have the config file in place, you can run the bundler. Run the following command in the root directory of your project:

<Tabs groupId="package-manager" queryString defaultValue={constants.defaultPackageManager} values={constants.packageManagers}>
<TabItem value="npm">

```shell
npm start
```

</TabItem>
<TabItem value="yarn">

```shell
yarn start
```

</TabItem>
</Tabs>

Now build and run your Android app as normal.

Once you reach your React-powered Activity inside the app, it should load the JavaScript code from the development server and display:

![Screenshot](/docs/assets/EmbeddedAppIOSVideo.gif)

> ### Creating a release build in Android Studio

Reviewer: ```suggestion
### Creating a release build in Xcode
```

You can use Xcode to create your release builds too! The only additional step is to add a script that is executed when the app is built to package your JS and images into the iOS application.

> 1. In Xocde, select your application

Reviewer 2: ```suggestion
1. In Xcode, select your application
```

This guide also assumes you're familiar with the basics of iOS development such as creating a `UIViewController` and editing the `Podfile` file.

### 1. Set up directory structure

To ensure a smooth experience, create a new folder for your integrated React Native project, then **move your existing iOS project** to the `/ios` subfolder.

## 2. Install NPM dependencies

Go to the root directory and run the following command:

```
wget https://raw.githubusercontent.com/react-native-community/template/refs/heads/0.76-stable/template/package.json
```

This will copy the `package.json` [file from the Community template](https://github.com/react-native-community/template/blob/0.76-stable/template/package.json) to your project.

Next, install the NPM packages by running:

<Tabs groupId="package-manager" queryString defaultValue={constants.defaultPackageManager} values={constants.packageManagers}>
<TabItem value="npm">

```shell
npm install
```

</TabItem>
<TabItem value="yarn">

```shell
yarn install
```

</TabItem>
</Tabs>

> Installation process has created a new `/node_modules` folder. This folder stores all the JavaScript dependencies required to build your project.

Reviewer 3: ```suggestion
Installation process has created a new `node_modules` folder. This folder stores all the JavaScript dependencies required to build your project.
```

Reviewer 3: Yes please 🙏

Add `node_modules/` to your `.gitignore` file (here the [Community default one](https://github.com/react-native-community/template/blob/0.76-stable/template/_gitignore)).

### 3. Install Development tools

### Command Line Tools for Xcode

Install the Command Line Tools. Choose **Settings... (or Preferences...)** in the Xcode menu. Go to the Locations panel and install the tools by selecting the most recent version in the Command Line Tools dropdown.

![Xcode Command Line Tools](/docs/assets/GettingStartedXcodeCommandLineTools.png)

### CocoaPods

[CocoaPods](https://cocoapods.org) is a package management tool for iOS and macOS development. We use it to add the actual React Native framework code locally into your current project.

> We recommend installing CocoaPods using [Homebrew](https://brew.sh/).

Reviewer 3: ```suggestion
We recommend installing CocoaPods using [Homebrew](https://brew.sh/):
```

```shell
brew install cocoapods
```

> > It is technically possible not to use CocoaPods, but that would require manual library and linker additions that would overly complicate this process.

Reviewer 3: This adds nothing to this page. I would just remove it

## 4. Adding React Native to your app

### Configuring CocoaPods

To configure CocoaPods, we need two files:

- A **Gemfile** that defines which Ruby dependencies we need.
- A **Podfile** that defines how to properly install our dependencies.

For the **Gemfile**, go to the root directory of your project and run this command

```sh

> wget https://raw.githubusercontent.com/react-native-community/template/refs/heads/0.76-stable/template/Gemfile

Author: Q: should we update the Android guide too then?

```

This will download the Gemfile from the template.
Similarly, for the **Podfile**, go to the `ios` folder of your project and run

```sh
wget https://raw.githubusercontent.com/react-native-community/template/refs/heads/0.76-stable/template/ios/Podfile
```

Please use the Community Template as a reference point for the [Gemfile](https://github.com/react-native-community/template/blob/0.76-stable/template/Gemfile) and for the [Podfile](https://github.com/react-native-community/template/blob/0.76-stable/template/ios/Podfile).

:::note
Remember to change [this line](https://github.com/react-native-community/template/blob/0.76-stable/template/ios/Podfile#L17) and [this line](https://github.com/react-native-community/template/blob/0.76-stable/template/ios/Podfile#L26) of the Podfile to match the name of your app.

> If your app don't have tests, rememebr to remove [this block](https://github.com/react-native-community/template/blob/0.76-stable/template/ios/Podfile#L26-L29).

Reviewer 3: ```suggestion
If your app don't have tests, remember to remove [this block](https://github.com/react-native-community/template/blob/0.76-stable/template/ios/Podfile#L26-L29).
```

:::

Now, we need to run a couple of extra commands to install the Ruby gems and the Pods.
Navigate to the `ios` folder and run the following commands:

```sh
bundle install
bundle exec pod install
```

The first command will install the ruby dependencies and the second command will actually integrate the React Native code in your application so that your iOS files can import the React Native headers.

## 5. Writing the TypeScript Code

Now we will actually modify the native Android application to integrate React Native.

> The first bit of code we will write is the actual React Native code for the new screen that will be integrated into our application.

Author: My personal interpretation is that a screen is the whole screen. A surface could be just a view or a set of views that occupy part of the screen.

        contentInsetAdjustmentBehavior="automatic"
        style={backgroundStyle}>
        <Header />
        <View
          style={{
            backgroundColor: isDarkMode
              ? Colors.black
              : Colors.white,
            padding: 24,
          }}>
          <Text style={styles.title}>Step One</Text>
          <Text>
            Edit <Text style={styles.bold}>App.tsx</Text> to
            change this screen and see your edits.
          </Text>
          <Text style={styles.title}>See your changes</Text>
          <ReloadInstructions />
          <Text style={styles.title}>Debug</Text>
          <DebugInstructions />
        </View>
      </ScrollView>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  title: {
    fontSize: 24,
    fontWeight: '600',
  },
  bold: {
    fontWeight: '700',
  },
});

export default App;
```

Here the [Community template file as reference](https://github.com/react-native-community/template/blob/0.76-stable/template/App.tsx)

## 5. Integrating with your iOS code

We now need to add some native code in order to start the React Native runtime and tell it to render our React components.

### Requirements

> 

Reviewer 3: > but we are getting closer to it

If you have a GH issue to link, please add it. Otherwise I would remove this reference

> React Native is supposed to work with the AppDelegate. Sadly it does not supports the SceneDelegate pattern as of today, but we are getting closer to it.

Reviewer 3: ```suggestion
React Native is supposed to work with the `AppDelegate`. Sadly it does not supports the `SceneDelegate` pattern as of today, but we are getting closer to it.
```

Reviewer 2: Can we link to docs for AppDelegate and SceneDelegate?

> The following part assumes that you don't have a SceneDelegate and that your application AppDelegate looks like this:

Reviewer 3: ```suggestion
The following part assumes that your `AppDelegate` looks like this:
```

<Tabs groupId="ios-language" queryString defaultValue={constants.defaultAppleLanguage} values={constants.appleLanguages}>
<TabItem value="objc">

```objc title="AppDelegate.m"
#import "AppDelegate.h"
#import "ViewController.h"

@interface AppDelegate ()

@end

@implementation AppDelegate {
  UIWindow *window;
}

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
  window = [UIWindow new];
  window.rootViewController = [ViewController new];
  [window makeKeyAndVisible];
  return YES;
}

@end
```

</TabItem>
<TabItem value="swift">

```swift title="AppDelegate.swift"
import UIKit

@main
class AppDelegate: UIResponder, UIApplicationDelegate {

  var window: UIWindow?

  func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
    // Override point for customization after application launch.
    window = UIWindow()
    window?.rootViewController = ViewController()
    window?.makeKeyAndVisible()
    return true
  }
}
```

</TabItem>
</Tabs>

### Update the AppDelegate class

> First, we need to extends the AppDelegate to inherit from one of the classes provided by React Native: `RCTAppDelegate`.

Reviewer 3: ```suggestion
First, we need to extends the `AppDelegate` to inherit from one of the classes provided by React Native: `RCTAppDelegate`.
```

3. The methods `sourceURLForBridge(for:)` and `bundleURL()` are used by the App to tell to React Native wher it can find the JS bundle that needs to be rendered. The `sourceURLForBridge(for:)` is from the Old Architecture and you can see that it is deferring the decision to the `bundleURL()` method, required by the New Architecture.

</TabItem>
</Tabs>

#### Presenting a React Native view in a rootViewController

Finally, we can present our React Native view. To do so, we need a new View Controller that can host a view in which we can load the JS content.

1. From Xcode, let's create a new `UIViewController` (Let's call it `ReactViewController`).
2. Have the Initial `ViewController` present the `ReactViewController`. There are several ways to do so, depending on your app. For this example, we assume that you have a button that presents React Native Modally.

<Tabs groupId="ios-language" queryString defaultValue={constants.defaultAppleLanguage} values={constants.appleLanguages}>
<TabItem value="objc">

```diff title="ViewController.m"
#import "ViewController.h"
+#import "ReactViewController.h"

@interface ViewController ()

@end

@implementation ViewController

> - (void)viewDidLoad {

Reviewer 3: Is this correct with the `-`?

Author: yes, in objc this denotes an instance method. a `+` instead, denotes a static method. Yes, the language is weird.

+    ])
  }
}
```

</TabItem>
</Tabs>

3. Update the `ReactViewController` code as it follows:

<Tabs groupId="ios-language" queryString defaultValue={constants.defaultAppleLanguage} values={constants.appleLanguages}>
<TabItem value="objc">

```diff title="ReactViewController.m"
#import "ReactViewController.h"
+#import <React-RCTAppDelegate/RCTRootViewFactory.h>
+#import <React-RCTAppDelegate/RCTAppDelegate.h>

@interface ReactViewController ()

@end

@implementation ReactViewController

 - (void)viewDidLoad {
   [super viewDidLoad];
   // Do any additional setup after loading the view.
+   RCTRootViewFactory * factory = ((RCTAppDelegate *)RCTSharedApplication().delegate).rootViewFactory;
+   self.view = [factory viewWithModuleName:@"HelloWorld"];
 }

@end
```

</TabItem>
<TabItem value="swift">

```diff title="ReactViewController.swift"
import UIKit
+import React_RCTAppDelegate

class ReactViewController: UIViewController {

  override func viewDidLoad() {
    super.viewDidLoad()

+    let factory = (RCTSharedApplication()?.delegate as? RCTAppDelegate)?.rootViewFactory;
+    self.view = factory?.view(withModuleName: "HelloWorld");
  }
}
```

</TabItem>
</Tabs>

> 4. Make sure to disable the Sandbox scripting. To achieve this, in Xcode, click on your app, then on build settings. Filter for script and set the `User Script Sandboxing` to `NO`.

Reviewer 2: Can you add a sentence about why you disable Sandbox scripting?  If it isn't important please ignore, just the first question that popped into my head.

Author: Sure, it's for Hermes, anyway.

import Tabs from '@theme/Tabs'; import TabItem from '@theme/TabItem'; import constants from '@site/core/TabsConstants';

## Key Concepts

> The keys to integrating React Native components into your iOS application are to:

Reviewer: Maybe we can also mention that this guide covers both old and new architecture? And if you want to use new arch you can install pods with a flag?

Author: this will be new arch only. All the references are to 0.76 and it will be published with that.
It follows the usual guidelines to disable it, eventually. But this is out of the scope of this guide, imho.

First, you need to create a `metro.config.js` file in the root of your project as follows:

```js
const {getDefaultConfig} = require('@react-native/metro-config');
module.exports = getDefaultConfig(__dirname);
```

You can checkout the [metro.config.js file](https://github.com/react-native-community/template/blob/0.76-stable/template/metro.config.js) from the Community template file as reference.

Once you have the config file in place, you can run the bundler. Run the following command in the root directory of your project:

<Tabs groupId="package-manager" queryString defaultValue={constants.defaultPackageManager} values={constants.packageManagers}>
<TabItem value="npm">

```shell
npm start
```

</TabItem>
<TabItem value="yarn">

```shell
yarn start
```

</TabItem>
</Tabs>

Now build and run your iOS app as normal.

Once you reach your React-powered Activity inside the app, it should load the JavaScript code from the development server and display:

> ![Screenshot](/docs/assets/EmbeddedAppIOSVideo.gif)

Reviewer 2: Can we constrain the width (see [example code here](https://github.com/facebook/react-native-website/blob/61313ad2e9f756ea0c3e6dd87a1fd53e1f674d6e/docs/datepickerios.md?plain=1#L90))?

Currently it's like this:
![CleanShot 2024-10-02 at 12 05 52@2x](https://github.com/user-attachments/assets/91ebdb9a-0e02-42b1-abff-3da2c04d6845)

If I set a max with of 300px it looks like this:
![CleanShot 2024-10-02 at 12 06 28@2x](https://github.com/user-attachments/assets/3b65df8c-9ab3-42cb-a875-ab486a3b491c)

Author: we need to do it on android as well.
